proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+
+```swift
+borrowing view: Span<Element> = ...
+for borrowing element in view {
+  doSomething(element)
+}
+```
+
+In the meantime, it is possible to loop through a `Span`'s elements by direct indexing:
+
+```swift
+func doSomething(_ e: borrowing Element) { ... }
+let view: Span<Element> = ...
+// either:
+var i = 0
+while i < view.count {
+  doSomething(view[i])
+  view.index(after: &i)
+}
+
+// ...or:
+for i in 0..<view.indices {
+  doSomething(view[i])
+}
+```
+
+##### Collection-like protocols for non-copyable and non-escapable types
+
+Non-copyable and non-escapable containers would benefit from a `Collection`-like protocol family to represent a set basic, common operations. This may be `Collection` if we find a way to make it work; it may be something else.
+
+Alongside this work, it may make sense to add a `Span` alternative to `withContiguousStorageIfAvailable()`, `RawSpan` alternative to `withUnsafeBytes`, etc., and seek to deprecate any closure-based API around unsafe pointers.
+
+##### Sharing piecewise-contiguous memory
+
+Some types store their internal representation in a piecewise-contiguous manner, such as trees and ropes. Some operations naturally return information in a piecewise-contiguous manner, such as network operations. These could supply results by iterating through a list of contiguous chunks of memory.
+
+##### Safe mutations of memory with `MutableSpan<T>`
+
+Some data structures can delegate mutations of their owned memory. In the standard library we have `withMutableBufferPointer()`, for example.
+
+The `UnsafeMutableBufferPointer` passed to a `withUnsafeMutableXXX` closure-style API is unsafe in multiple ways:
+
+1. The pointer itself is unsafe and unmanaged
+2. `subscript` is only bounds-checked in debug builds of client code
+3. It might escape the duration of the closure
+4. Exclusivity of writes is not enforced
+5. Initialization of any particular memory address is not ensured
+
+I.e., it is unsafe in all the ways `UnsafeBufferPointer`-passing closure APIs are unsafe in addition to being unsafe in exclusivity and in initialization.
+
+Loading an uninitialized non-`BitwiseCopyable` value leads to undefined behavior. Loading an uninitialized `BitwiseCopyable` value does not immediately lead to undefined behavior, but it produces a garbage value which may lead to misbehavior of the program.
+
+A `MutableSpan<T>` should provide a better, safer alternative to mutable memory in the same way that `Span<T>` provides a better, safer read-only type. `MutableSpan<T>` would also automatically enforce exclusivity of writes.
+
+However, it alone does not track initialization state of each address, and that will continue to be the responsibility of the developer.
+
+##### Delegating initialization of memory with `OutputSpan<T>`
+
```

Comment:
I'll update this with the current naming around `OutputSpan`.

Reply:
This section also needs to talk about the simple case where there is always an initialized prefix with an uninitialized tail. (will do)

Reply:
I've improved this section.